FreeRTOS/FreeRTOS-Kernel
# Add VeriFast thread safety proof for context switches in SMP port

Add VeriFast thread safety proof for context switches in SMP port

Description
-----------
The SMP implementation of the scheduler suffers from a buffer underflow, cf. the fix proposed in [PR 607](https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/607). We used [VeriFast](https://github.com/verifast/verifast), a deductive program verifier for C, to prove that once the fix from PR 607 has been applied, context switches are memory safe (i.e. no memory error) and mutually thread safe (i.e. no race condition). 

The proof is an *unbounded* proof. That is, it considers any possible number of tasks and any possible sizes for the involved data structures. Further, it considers any possible task interleavings and any possible interrupt schedule that may occur during runtime. Taking all this into account, it ensure that no memory error and no race condition can occur during runtime, when we run multiple concurrent context switches on different cores.

See the proof README for a more detailed explanation.

Test Steps
-----------
Run one of the scripts `run-verifast.sh` or `run-vfide.sh` as descripted in the proof README.

Related Issue
-----------
https://github.com/FreeRTOS/FreeRTOS-Kernel/pull/607


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
